diff -ur a/FL/filename.H b/FL/filename.H
--- a/FL/filename.H	2010-12-20 13:35:43.000000000 +0000
+++ b/FL/filename.H	2012-08-24 19:50:30.000000000 +0100
@@ -79,7 +79,7 @@
 #  endif /* __cplusplus */
 
 
-#  if defined(WIN32) && !defined(__CYGWIN__) && !defined(__WATCOMC__)
+#  if 0
 
 struct dirent {char d_name[1];};
 
diff -ur a/Makefile b/Makefile
--- a/Makefile	2011-02-15 15:29:03.000000000 +0000
+++ b/Makefile	2012-08-24 19:49:07.000000000 +0100
@@ -27,7 +27,7 @@
 
 include makeinclude
 
-DIRS = $(IMAGEDIRS) src $(CAIRODIR) fluid test documentation
+DIRS = $(IMAGEDIRS) src $(CAIRODIR) # fluid test documentation
 
 all: makeinclude fltk-config
 	for dir in $(DIRS); do\
